Don't unmount kubelet-root-dir on k0s reset

dockerpac opened the pull request:
## Description

Fixes an issue where `k0s reset` was unmounting and deleting the kubelet-root-dir mount point when configured on a separate filesystem.

When `--kubelet-root-dir` is on a dedicated partition (common in production), `k0s reset` would unmount the filesystem and delete the mount point.

This PR applies the same mount point preservation logic to `kubelet-root-dir` that was already existing for `data-dir`.


## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [ ] Auto test added

Tested scenarios:
- **Mounted filesystem**: Kubelet mount point preserved, only contents deleted
- **Default setup**: Directory completely removed as before

## Checklist

- [x] My code follows the style [guidelines](https://docs.k0sproject.io/head/contributors/) of this project
- [x] My commit messages are [signed-off](https://docs.k0sproject.io/head/contributors/github_workflow/)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

twz123 commented:
> When running the bootloose environment manually, I noticed that kubelet creates an expected mount:

I wouldn't know that kubelet is creating a mount for its data directory. This definitely needs some investigation, and to me, the integration test failures look legit, still.

vikramhh commented:
@dockerpac @ncopa @twz123 

By using bpftrace, it can be confirmed that the mounting of the configured value of kubelet root directory is done by kubelet itself. The failure above looks legit and arises not due to containerized environment per-se but likely due to the following sequence of calls in kubelet:

```
Kubelet.setupDataDirs()
  -> hostutil.MakeRShared(rootDir)
    -> DoMakeRShared(rootDir, /proc/self/mountinfo)
      -> if not already shared:
           bind-mount rootDir onto itself
           make it recursively shared
```

The above is the kubelet:

1. Needing its root dir mount tree to support mount propagation
2. Checking whether the relevant mount is already shared
3. If not, it making the kubelet root dir a mountpoint by doing a self bind mount

By doing the following before installing k0s, you can force this behavior on a "real linux box" as well - it was confirmed on aws ec2 ubuntu[ the following assumes that `/var/lib/kubelet` is the configured value of kubelet root directory]:

```
sudo mkdir -p /var/lib/kubelet
sudo mount --make-private /

```

The reverse (making the root non-private/shared on bootloose) was not tried to make the failure go away because IMO in this instance it is the test that needs to be modified for reasons like scoping things tightly and because the "failure" actually reflects what should be happening here with this PR-s changes. Besides bootloose runs the node as a container and it would be somewhat unexpected for the container's root to have shared propagation.

If you concur, I can push the test changes otherwise could you suggest what else we could try?

jnummelin commented:
> Besides bootloose runs the node as a container and it would be somewhat unexpected for the container's root to have shared propagation

We mitigate this in bootloose based tests by mounting `/var/lib/k0s` as a volume. So the tests involved with custom kubelet dir could maybe do the same by attaching an extra volume for the given dir. Something like:

```
func TestXYZSuite(t *testing.T) {
	s := XYZSuite{
		common.BootlooseSuite{
			ControllerCount: 1,
			WorkerCount:     1,
			ExtraVolumes: []config.Volume{
				{
					Type:        "volume",
					Destination: "/var/lib/kubelet",
				},
			},
		},
	}
	suite.Run(t, &s)
}
```

vikramhh commented:
Hi @jnummelin - the CI workflows are awaiting your approval from the fork. Thanks!

vikramhh committed:
Don't unmount a kubelet root dir that k0s didn't mount on reset

When the kubelet root dir is on its own mount/volume (e.g. --kubelet-root-dir
points at a dedicated volume), k0s reset unmounted that mount -- which k0s
never created -- and then aborted because it could not remove the now-busy
mount point ("failed to delete k0s kubelet root directory: ... device or
resource busy").

Treat such a kubelet root dir like the data dir: leave its own mount in place
and only empty its contents, tolerating the unlinkat error on the busy mount
point. Mounts nested under the kubelet root dir are still unmounted, and a
kubelet root dir that lives under the data dir is still unmounted and removed
together with the data dir.

Expand the reset smoke test to cover both a kubelet root dir on its own volume
(kept and emptied) and one under the data dir (removed with it), to assert the
kubelet root dir is emptied, and to verify a mount nested under the kubelet
root dir is unmounted and removed while its out-of-tree bind source survives.
Also fix a typo in the error message.

Signed-off-by: vikramhh <vsingh@mirantis.com>

vikramhh commented:
Hi @jnummelin - could you please approval a workflow run for the latest push. Thanks!

vikramhh commented:
@twz123 @makhov 

The two failing checks — `check-dualstack-kuberouter` and `check-dualstack-kuberouter-dynamicconfig` — fail on the same assertion, and it's unrelated to this PR:

`dualstack_test.go:252: pod metrics-server-d987bf784-xxxxx container metrics-server has 1 restarts. Should be 0
`

This is `VerifyNoRestartedPods` tripping because the metrics-server container restarted once during cluster bring-up — a transient flake. (The No agent available lines earlier in the log are just retries that subsequently succeed — "Connection to ipv4 address ... was successful" — not the cause.)

This PR only touches `pkg/cleanup/directories.go` (the k0s reset path) and `inttest/reset/reset_test.go`. The dualstack suite never runs k0s reset, so the changed code path isn't exercised by these tests at all.

The same `metrics-server-restart` failure has recurred on several unrelated PRs over the last few days — dependency bumps and networking changes that can't be affected by reset/cleanup logic — which rules this change out as the cause.

Could you please re-run the two failed jobs? Everything else (108 checks) is green. Thanks!

(There doesn't appear to be a tracking issue for this particular dualstack flake yet — happy to open one if that's useful.)

vikramhh commented:
Hey @twz123 @makhov @jnummelin - tests runs are all green, could you please take a look.
